File: medium/Clone_Graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if node is None:
            return None
        def next_node(node, adj_node):
            for neighbor in node.neighbors:
                if neighbor in val_dict:
                    if val_dict[neighbor] in adj_node.neighbors:
                        break
                    adj_node.neighbors.append(val_dict[neighbor])
                else:
                    val_dict[neighbor] = Node(neighbor.val, [])
                    adj_node.neighbors.append(val_dict[neighbor])
                next_node(neighbor, adj_node.neighbors[-1])
        val_dict = {}
        ans = Node(node.val, [])
        val_dict[node] = ans
        next_node(node, ans)
        return ans
        
        
        
        
# """
# # Definition for a Node.
# class Node:
#     def __init__(self, val = 0, neighbors = None):
#         self.val = val
#         self.neighbors = neighbors if neighbors is not None else []
# """

# class Solution:
#     def __init__(self):
#         self.visited = {}
        
#     def cloneGraph(self, node: 'Node') -> 'Node':
#         if not node:
#             return node     # None
#         if node in self.visited:
#             return self.visited[node]
        
#         clone = Node(node.val, [])
#         self.visited[node] = clone
        
#         if node.neighbors:
#             for n in node.neighbors:
#                 clone.neighbors.append(self.cloneGraph(n))
            
#         return clone

File: cpp/medium/Clone_Graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return nullptr;
        queue<Node*> q;
        unordered_map<int, Node*> already_map;
        Node* ans_node = new Node(node->val);
        already_map[ans_node->val] = ans_node;
        q.push(node);
        while (!q.empty()) {
            Node* n = q.front();
            q.pop();
            Node* new_n_node = already_map.find(n->val)->second;
            for (auto neighbor:n->neighbors) {
                if (already_map.find(neighbor->val) == already_map.end()) {
                    Node* new_node = new Node(neighbor->val);
                    already_map[new_node->val] = new_node;
                    q.push(neighbor);
                }
                new_n_node->neighbors.push_back(already_map.find(neighbor->val)->second);
            }
        }
        return ans_node;
    }
};
